[visionOS] `ensurePositionInformationIsUpToDate` is called every time a scroll gesture occurs

lilyspiniolas opened the pull request:
#### f08f67b1112e2308cff608b60fe6d3de1207a441
<pre>
[visionOS] `ensurePositionInformationIsUpToDate` is called every time a scroll gesture occurs
<a href="https://bugs.webkit.org/show_bug.cgi?id=300964">https://bugs.webkit.org/show_bug.cgi?id=300964</a>
<a href="https://rdar.apple.com/157183410">rdar://157183410</a>

Reviewed by Wenson Hsieh.

In visionOS 26, support for the &lt;model&gt; element was added, which includes a new pan gesture recognizer
for the purpose of rotating model elements. This new recognizer means that pinch to scroll gestures are
first recognized as a potential pan gesture, and to determine whether we should actually pan, we check
if the gesture is occuring over a model element. However, in order to do this, we have to call
`ensurePositionInformationIsUpToDate` in WKContentViewInteraction.

`ensurePositionInformationIsUpToDate` must fetch the information from the web process if valid information
is not already present, and since this occurs synchronously, it can cause hangs. Since it is now called
for every scroll gesture, the number of hangs has increased.

To mitigate this, we now track whether or not a model element is present on the page. If one is present,
we continue as we did before, and if one is not present, we immediately stop processing the gesture as
a pan gesture, avoiding the unnecessary call to `ensurePositionInformationIsUpToDate`.

To track the presence of a model element, each Document and Page now keep a count of their total number
of model elements. When an HTMLModelElement is inserted or removed, it notifies the document that the
count should increase. The document then notifies the Page that the count should increase. When the count
changes to 0 or 1, the Page notifies the UI process that we either do or do not have a model element.

During document teardown, HTMLModelElement::removedFromAncestor is called after the page has already
been disconnected, which means the Page has an inaccurate model element count unless it is notified
in another manner. To achieve this, we decrement the count in Document::destroyRenderTree so that
the page can still be notified.

Added multiple layout tests to ensure that `ensurePositionInformationIsUpToDate` is only called during
scroll gestures if a model element is present on the page. Added UIHelper methods
`didCallEnsurePositionInformationIsUpToDateSinceLastCheck` and
`clearEnsurePositionInformationIsUpToDateTracking` to achieve this.

Tests: model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames.html
       model-element/model-element-determines-if-scroll-gesture-updates-position-information-navigation.html
       model-element/model-element-determines-if-scroll-gesture-updates-position-information.html
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-expected.txt: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames-expected.txt: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames.html: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-navigation-expected.txt: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-navigation.html: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information.html: Added.
* LayoutTests/model-element/resources/finish-models-navigation.html: Added.
* LayoutTests/model-element/resources/models-frame.html: Added.
* LayoutTests/model-element/resources/no-models.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.didCallEnsurePositionInformationIsUpToDateSinceLastCheck.return.new.Promise):
(window.UIHelper.didCallEnsurePositionInformationIsUpToDateSinceLastCheck):
(window.UIHelper.clearEnsurePositionInformationIsUpToDateTracking.return.new.Promise):
(window.UIHelper.clearEnsurePositionInformationIsUpToDateTracking):
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::destroyRenderTree):
(WebCore::Document::incrementModelElementCount):
(WebCore::Document::decrementModelElementCount):
* Source/WebCore/dom/Document.h:
(WebCore::Document::hasModelElement const):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setHasModelElement):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::incrementModelElementCount):
(WebCore::Page::decrementModelElementCount):
* Source/WebCore/page/Page.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView didEnsurePositionInformationIsUpToDate]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setHasModelElement):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView ensurePositionInformationIsUpToDate:]):
(-[WKContentView gestureRecognizerShouldBegin:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setHasModelElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setHasModelElement):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::didCallEnsurePositionInformationIsUpToDateSinceLastCheck const):
(WTR::UIScriptController::clearEnsurePositionInformationIsUpToDateTracking):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView didEnsurePositionInformationIsUpToDate]):
(-[TestRunnerWKWebView didCallEnsurePositionInformationIsUpToDateSinceLastCheck]):
(-[TestRunnerWKWebView clearEnsurePositionInformationIsUpToDateTracking]):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::didCallEnsurePositionInformationIsUpToDateSinceLastCheck const):
(WTR::UIScriptControllerIOS::clearEnsurePositionInformationIsUpToDateTracking):

Canonical link: <a href="https://commits.webkit.org/301766@main">https://commits.webkit.org/301766@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bdd6be78cbb351e304beec17fb5a1019be24fd8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126970 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46606 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133974 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78540 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9085998e-74e5-47af-b9d2-912d77781bf6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128841 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55134 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96619 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64615 "") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8bcdd4e9-1605-41e5-8814-e6f748a5d28b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37790 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113571 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77130 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e0a18214-5f30-4580-8caf-b28ed8933fc3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36652 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31733 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77367 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107628 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32058 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136499 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53626 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41292 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105133 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54129 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109935 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104824 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26733 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50342 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28686 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51110 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53558 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59430 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52798 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56131 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54557 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -1045,6 +1047,9 @@ FOR_EACH_PRIVATE_WKCONTENTVIEW_ACTION(DECLARE_WKCONTENTVIEW_ACTION_FOR_WEB_VIEW)
 #endif
 
+- (BOOL)_didCallEnsurePositionInformationIsUpToDateSinceLastCheck;
+- (void)_clearEnsurePositionInformationIsUpToDateTracking;
```
Minor nit — it _may_ be a bit less intrusive to add a `WKWebView` method hook along the lines of `-didEnsurePositionInformationIsUpToDate` (similar to `-didStartFormControlInteraction`), and have the test runner control the logic to override the method and increment a counter (or manage a boolean flag) whenever the method is called.

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -3279,6 +3279,7 @@ - (void)doAfterPositionInformationUpdate:(void (^)(WebKit::InteractionInformatio
 
 - (BOOL)ensurePositionInformationIsUpToDate:(WebKit::InteractionInformationRequest)request
 {
+    _didCallEnsurePositionInformationIsUpToDateSinceLastCheck = YES;
```
(I also wonder if we should only note this method call, in the case where sync IPC is actually being sent…)

lilyspiniolas replied:
Since the issue is that `ensurePositionInformationIsUpToDate` is being called at all when it shouldn't be, I think it makes more sense to keep as-is for testing purposes, even if it might not necessarily indicate sync IPC is being sent.

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3692,6 +3697,10 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool m_registeredForFullSpeedUpdates { false };
 
+#if ENABLE(MODEL_ELEMENT)
+    bool m_hasModelElement { false };
```
Do we also need to reset this state in `WebPageProxy::resetStateAfterProcessExited`?

whsieh commented on Source/WebCore/dom/Document.h:
```diff
@@ -2306,6 +2309,7 @@ class Document
     std::unique_ptr<LazyLoadImageObserver> m_lazyLoadImageObserver;
 #if ENABLE(MODEL_ELEMENT)
     std::unique_ptr<LazyLoadModelObserver> m_lazyLoadModelObserver;
+    unsigned m_modelElementCount { 0 };
```
Hm... is it necessary to add this member to `Document`? (AFAICT it's only used for a debug assertion — if so, perhaps we could limit it to debug builds only)

lilyspiniolas replied:
The precise number of model elements needs to be tracked in order to decrement the counter by `m_modelElementCount` in `Document::destroyRenderTree()`

pxlcoder commented on LayoutTests/model-element/resources/finish-models-navigation.html:
```diff
@@ -0,0 +1,71 @@
+</head>
+<script>
+    jsTestIsAsync = true;
+    testRunner.waitUntilDone();
```
No need for `testRunner.waitUntilDone` since you have `jsTestIsAsync`.

pxlcoder commented on LayoutTests/model-element/resources/finish-models-navigation.html:
```diff
@@ -0,0 +1,71 @@
+                attempting to scroll, ensurePositionInformationIsUpToDate should not be called.`);
+
+            await scrollAndCheckIfPositionInformationUpdated(false);
+            testRunner.notifyDone();
```
Paired with `jsTestIsAsync`.
```suggestion
            finishJSTest();
```

pxlcoder commented on LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames.html:
```diff
@@ -0,0 +1,155 @@
+</head>
+<script>
+    jsTestIsAsync = true;
+    testRunner.waitUntilDone();
```
Same comments as above.

pxlcoder commented on LayoutTests/model-element/resources/finish-models-navigation.html:
```diff
@@ -0,0 +1,71 @@
+<script>
+    jsTestIsAsync = true;
+    testRunner.waitUntilDone();
+    testRunner.dumpAsText();
```
I don't think this is necessary either.

whsieh commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11407,6 +11416,31 @@ LazyLoadModelObserver& Document::lazyLoadModelObserver()
+void Document::decrementModelElementCount()
+{
+    RefPtr page = this->page();
+    if (!page)
```
Is it possible that models could get under- or over-counted if the document is disconnected from the frame, or the frame is detached (and we therefore hit this early return only in increment or decrement)? (I would double check navigating forward/back with model elements, but _perhaps_ this is fine…)

lilyspiniolas replied:
In `Document::destroyRenderTree()` we subtract all tracked model elements from the Page's count so that the count remains correct after the document is disconnected from the page, does that seem suitable?

whsieh replied:
Oh, I'd totally missed that; I see.

And for the `insertedIntoAncestor` part, it also seems like we already assume `document().page()` must be non-null already so that part is probably fine!

mwyrzykowski commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -3468,6 +3469,9 @@ - (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
 
 #if ENABLE(MODEL_PROCESS)
     if (gestureRecognizer == _modelInteractionPanGestureRecognizer) {
+        if (!_page->hasModelElement())
```
It looks like `hasModelElement()` is only used when `ENABLE(MODEL_PROCESS)` is `1` but defined when `ENABLE(MODEL_ELEMENT)` is `1`.

Should it only be defined  when `ENABLE(MODEL_PROCESS)` is `1`?

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -742,6 +742,8 @@ typedef NS_OPTIONS(NSUInteger, _WKWebViewDataType) {
 - (void)didEndFormControlInteraction WK_API_AVAILABLE(ios(10.3));
 
+- (void)didEnsurePositionInformationIsUpToDate WK_API_AVAILABLE(ios(26.0), visionos(26.0));
+
```
Oh, one minor nit — it should be `WK_*_TBA` here in the availability. Or alternately, you can put it in `WKWebViewPrivateForTestingIOS.h`.

(Ideally, both of the above methods should also be moved here, though they were introduced before `WKWebViewPrivateForTestingIOS.h` existed)

lilyspiniolas committed:
[visionOS] `ensurePositionInformationIsUpToDate` is called every time a scroll gesture occurs
https://bugs.webkit.org/show_bug.cgi?id=300964
rdar://157183410

Reviewed by Wenson Hsieh.

In visionOS 26, support for the <model> element was added, which includes a new pan gesture recognizer
for the purpose of rotating model elements. This new recognizer means that pinch to scroll gestures are
first recognized as a potential pan gesture, and to determine whether we should actually pan, we check
if the gesture is occuring over a model element. However, in order to do this, we have to call
`ensurePositionInformationIsUpToDate` in WKContentViewInteraction.

`ensurePositionInformationIsUpToDate` must fetch the information from the web process if valid information
is not already present, and since this occurs synchronously, it can cause hangs. Since it is now called
for every scroll gesture, the number of hangs has increased.

To mitigate this, we now track whether or not a model element is present on the page. If one is present,
we continue as we did before, and if one is not present, we immediately stop processing the gesture as
a pan gesture, avoiding the unnecessary call to `ensurePositionInformationIsUpToDate`.

To track the presence of a model element, each Document and Page now keep a count of their total number
of model elements. When an HTMLModelElement is inserted or removed, it notifies the document that the
count should increase. The document then notifies the Page that the count should increase. When the count
changes to 0 or 1, the Page notifies the UI process that we either do or do not have a model element.

During document teardown, HTMLModelElement::removedFromAncestor is called after the page has already
been disconnected, which means the Page has an inaccurate model element count unless it is notified
in another manner. To achieve this, we decrement the count in Document::destroyRenderTree so that
the page can still be notified.

Added multiple layout tests to ensure that `ensurePositionInformationIsUpToDate` is only called during
scroll gestures if a model element is present on the page. Added UIHelper methods
`didCallEnsurePositionInformationIsUpToDateSinceLastCheck` and
`clearEnsurePositionInformationIsUpToDateTracking` to achieve this.

Tests: model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames.html
       model-element/model-element-determines-if-scroll-gesture-updates-position-information-navigation.html
       model-element/model-element-determines-if-scroll-gesture-updates-position-information.html
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-expected.txt: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames-expected.txt: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-frames.html: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-navigation-expected.txt: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information-navigation.html: Added.
* LayoutTests/model-element/model-element-determines-if-scroll-gesture-updates-position-information.html: Added.
* LayoutTests/model-element/resources/finish-models-navigation.html: Added.
* LayoutTests/model-element/resources/models-frame.html: Added.
* LayoutTests/model-element/resources/no-models.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.didCallEnsurePositionInformationIsUpToDateSinceLastCheck.return.new.Promise):
(window.UIHelper.didCallEnsurePositionInformationIsUpToDateSinceLastCheck):
(window.UIHelper.clearEnsurePositionInformationIsUpToDateTracking.return.new.Promise):
(window.UIHelper.clearEnsurePositionInformationIsUpToDateTracking):
* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::destroyRenderTree):
(WebCore::Document::incrementModelElementCount):
(WebCore::Document::decrementModelElementCount):
* Source/WebCore/dom/Document.h:
(WebCore::Document::hasModelElement const):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setHasModelElement):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::incrementModelElementCount):
(WebCore::Page::decrementModelElementCount):
* Source/WebCore/page/Page.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView didEnsurePositionInformationIsUpToDate]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setHasModelElement):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView ensurePositionInformationIsUpToDate:]):
(-[WKContentView gestureRecognizerShouldBegin:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setHasModelElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setHasModelElement):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::didCallEnsurePositionInformationIsUpToDateSinceLastCheck const):
(WTR::UIScriptController::clearEnsurePositionInformationIsUpToDateTracking):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView didEnsurePositionInformationIsUpToDate]):
(-[TestRunnerWKWebView didCallEnsurePositionInformationIsUpToDateSinceLastCheck]):
(-[TestRunnerWKWebView clearEnsurePositionInformationIsUpToDateTracking]):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::didCallEnsurePositionInformationIsUpToDateSinceLastCheck const):
(WTR::UIScriptControllerIOS::clearEnsurePositionInformationIsUpToDateTracking):

Canonical link: https://commits.webkit.org/301766@main
